File: src/main/scala/com/github/shadowsocks/plugin/kcptun/BinaryProvider.scala
package com.github.shadowsocks.plugin.kcptun

import java.io.{File, FileNotFoundException}

import android.net.Uri
import android.os.ParcelFileDescriptor
import com.github.shadowsocks.plugin.{NativePluginProvider, PathProvider}

/**
  * @author Mygod
  */
final class BinaryProvider extends NativePluginProvider {
  override protected def populateFiles(provider: PathProvider): Unit = provider.addPath("kcptun", "755")

  override def getExecutable: String = getContext.getApplicationInfo.nativeLibraryDir + "/libkcptun.so"

  override def openFile(uri: Uri): ParcelFileDescriptor = uri.getPath match {
    case "/kcptun" => ParcelFileDescriptor.open(new File(getExecutable), ParcelFileDescriptor.MODE_READ_ONLY)
    case _ => throw new FileNotFoundException()
  }
}

File: app/src/main/java/com/github/shadowsocks/plugin/kcptun/BinaryProvider.kt
package com.github.shadowsocks.plugin.kcptun

import android.net.Uri
import android.os.ParcelFileDescriptor
import com.github.shadowsocks.plugin.NativePluginProvider
import com.github.shadowsocks.plugin.PathProvider
import java.io.File
import java.io.FileNotFoundException

class BinaryProvider : NativePluginProvider() {
    override fun populateFiles(provider: PathProvider) {
        provider.addPath("kcptun", 0b111101101)
    }
    override fun getExecutable() = context!!.applicationInfo.nativeLibraryDir + "/libkcptun.so"
    override fun openFile(uri: Uri?): ParcelFileDescriptor = when (uri?.path) {
        "/kcptun" -> ParcelFileDescriptor.open(File(getExecutable()), ParcelFileDescriptor.MODE_READ_ONLY)
        else -> throw FileNotFoundException()
    }
}
